SPEC.md

```diff
@@ -0,0 +1,258 @@
+
+<p align="center">
+  <br>
+  <img src="images/polling-stack.png" alt="Polling Fetcher Stack"/>
+</p>
+
+
+```js
+const pollingFetcher = new PollingFetcher(
+  10000, //Polling Interval in ms
+  new HTTPFetcher({ uri: 'http://api.githunt.com/graphql' })
+);
+```
+
+In addition to the query, operationName, and variables, `Operation` includes a context, which is modified and passed down the stack of fetchers. In addition, this context is sent to the server to through the query body. In this example, `context` is used by a polling fetcher to tell a caching fetcher whether a request should be returned from the cache or the network.
+
+<p align="center">
+  <br>
+  <img src="images/context.png" alt="Context Passing Stack Example"/>
+</p>
+
+## API
+
+The fetcher interface contains a single method:
+
+```js
+interface ApolloFetcher{
+  request(Operation) : Observable<FetchResult>
+}
+```
+
+An `Operation` contains all the information necessary to execute a GraphQL query, including the GraphQL AST:
+
+```js
+Operation {
+  query: DocumentNode,
+  operationName: string,
+  variables: object,
+  context: object,
+}
+```
+
+An `Observable` is returned from `request` and notifies a `Subscriber` on the life-cycle of a GraphQL request:
+
+```js
+Observable{
+  //Causes the underlying producer to start
+  subscribe(Subscriber<FetchResult>) : UnsubscribeHandler
+  subscribe(
+    next(FetcheResult) => void,
+    error(any) => void,
+    complete() => void) : UnsubscribeHandler
+}
+```
+
+A `Subscriber` is passed to `subscribe`
+
+```js
+Subscriber<T>{
+  next(T),
+  error(any),
+  complete(),
+}
+```
+
+`UnsubscribeHandler` is returned from `subscribe`
+
+```js
+type UnsubscribeHandler = () => void
+```
+
+`FetchResult` is passed to the `next` callback of `Subscriber`
+
+```js
+FetchResult{
+  data?: object,
+  errors?: object[],
+  extensions?: any,
+  context?:object,
+}
+```
+
+## More examples
+
+Here is a list of planned fetchers with selected diagrams:
+
+Base:
+
+* Websocket: subscriptions support
+* Batching-\*: support for batching operations with a transport
+
+<p align="center">
+  <br>
+  <img src="images/batch-fetcher.png" alt="Batch Fetcher"/>
+</p>
+
+Intermediate:
+
+* Polling: repeats requests on a specified interval
+
+<p align="center">
+  <br>
+  <img src="images/polling-fetcher.png" alt="Polling Fetcher"/>
+</p>
+
+* Caching: returns data if result in the cache and stores data in cache on response
+* Compose: combines a list of fetchers (would require additional semantics around constructor argument order)
+* Dedup: saves query signatures that are currently on the wire and returns the result for all of those queries
+* Retry: error callback causes the request to retry
+* Mock: returns fake data for all or part of a request ← implement with BaseFetcher (eg. executes `graphql`)
+
+Fork:
+
+* Split: split operations between fetchers depending on a function passed in
```

Comment:
I think writing out how the API for this could look would be really helpful in the overall design. Splitting is an interesting concept and may be possible simply by allowing fetchers to filter which requests they want to handle. 

```js
const XHR = HTTPFetcher({ uri: '...' })
  .filter(({ operation })=> isNotSubscription(operation))
const WS = WSFetcher({ socket: '...' })
  .filter(({ operation }) => isSubscription(operation))

const fetcher = Fetcher.from([XHR, WS]);
```